File: leetcode/queue_stack/3476_Maximize_Profit_from_Task_Assignment/codes/heap.py
from typing import List
import heapq
from collections import Counter

class Solution:
    def maxProfit(self, workers: List[int], tasks: List[List[int]]) -> int:
        res = 0
        additionalWorker = 1
        pq = [(-p, s) for s, p in tasks]
        heapq.heapify(pq)
        wrks = Counter(workers)  
        while pq:
            profit, skill = heapq.heappop(pq)
            profit = -profit
            if wrks[skill] > 0 or additionalWorker > 0:
                res += profit
                if wrks[skill] > 0:
                    wrks[skill] -= 1
                else:
                    additionalWorker = 0
        return res

File: leetcode/queue_stack/3476_Maximize_Profit_from_Task_Assignment/codes/heap.cpp
#include <vector>
#include <queue>
#include <unordered_map>
using namespace std;

class Solution {
public:
    long long maxProfit(vector<int>& workers, vector<vector<int>>& tasks) {
        long long res = 0;
        int additionalWorker = 1;
        priority_queue<pair<int,int>> pq; 
        for (auto& t : tasks) {
            pq.push({t[1], t[0]});
        }
        unordered_map<int,int> wrks;
        for (int w : workers) {
            wrks[w]++;
        }
        while (!pq.empty()) {
            auto [profit, skill] = pq.top();
            pq.pop();
            if (wrks[skill] > 0 || additionalWorker > 0) {
                res += profit;
                if (wrks[skill] > 0) {
                    wrks[skill]--;
                } else {
                    additionalWorker = 0;
                }
            }
        }
        return res;
    }
};
